common/transport/amqp/src/amqp.ts

```diff
@@ -145,44 +235,190 @@ export class Amqp {
           putToken: (audience, token, callback) => callback(new errors.NotConnectedError()),
         },
         connecting: {
-          _onEnter: (config, connectCallback) => {
-            let connectError = null;
-            const connectErrorHandler = (err) => {
-              connectError = err;
-            };
-
-            this._amqp.on(_amqpClientError, connectErrorHandler);
-            this._amqp.connect(config.uri, config.policyOverride)
-              .then((result) => {
-                debug('AMQP transport connected.');
-                this._amqp.on(_amqpClientError, amqpErrorHandler);
-                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
-                this._fsm.transition('connected', connectCallback, result);
-                return null;
-              })
-              .catch((err) => {
-                this._amqp.removeListener(_amqpClientError, connectErrorHandler);
-                /*Codes_SRS_NODE_COMMON_AMQP_16_003: [The `connect` method shall call the `done` callback if the connection fails.] */
-                this._fsm.transition('disconnected', connectCallback, connectError || err);
-              });
+          _onEnter: (connectionParameters, connectCallback) => {
+            this._rheaContainer.options.sender_options = {properties: {'com.microsoft:client-version': this._config.userAgentString}, reconnect: false};
+            this._rheaContainer.options.receiver_options = {properties: {'com.microsoft:client-version': this._config.userAgentString}, reconnect: false, autoaccept: autoSettleMode};
+            this._connectionCallback = connectCallback;
+            this._indicatedConnectionError = undefined;
+            this._disconnectionOccurred = false;
+            this._sessionCloseOccurred = false;
+            this._connectionCloseOccurred = false;
+            //
+            // According to the rhea maintainers, one can depend on that fact that no actual network activity
+            // will occur until the nextTick() after the call to connect.  Because of that, one can
+            // put the event handlers on the rhea connection object returned from the connect call and be assured
+            // that the listeners are in place BEFORE any possible events will be emitted on the connection.
+            //
+            this._rheaConnection = this._rheaContainer.connect(connectionParameters);
+            manageConnectionHandlers('on');
+          },
+          'connection_open': (context: EventContext) => {
+            debug('in the connection open handler for the connection state: ' + context.connection.name);
+            this._rheaConnection = context.connection;
+            let callback = this._connectionCallback;
```

Comment:
>let callback = this._connectionCallback; [](start = 12, length = 40)

Aaargh.  I've been stuck with this pattern before.  I hate saving callbacks like this.  It just seems so error prone.  So @#$&() frustrating.

Reply:
I was hoping there was some way to pass information via the EventContext, but I don't see anything easy.  Aargh.  My comment is just expressing frusturation.  

---
In reply to: [209376307](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209376307) [](ancestors = 209376307)

Reply:
Or less error prone.

---
In reply to: [209382324](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209382324) [](ancestors = 209382324,209376307)